database/database_manager.py

```diff
@@ -0,0 +1,93 @@
+"""
+database_manager.py - Debug and diagnostics commands
+Provides postgreSQL connectivity for mechasqueak3.
+Copyright (c) 2018 The Fuel Rat Mischief,
+All rights reserved.
+Licensed under the BSD 3-Clause License.
+See LICENSE.md
+"""
+from config import config
+from psycopg2 import sql, pool
+from utils.ratlib import Singleton
+import logging
+import psycopg2
+
+log = logging.getLogger(f"mecha.{__name__}")
+
+
+class DatabaseManager(Singleton, object):
+
+    def __init__(self,
+                 dbhost=None,
+                 dbport=None,
+                 dbname=None,
+                 dbuser=None,
+                 dbpassword=None
+                 ):
+
+        if not hasattr(self, "_initialized"):
+            self._initialized = True
+
+            self._dbhost = dbhost if dbhost is not None else config['database'].get('host')
+            assert self._dbhost
+
+            self._dbport = dbport if dbhost is not None else config['database'].get('port')
+            assert self._dbport
+
+            self._dbname = dbname if dbname is not None else config['database'].get('dbname')
+            assert self._dbname
+
+            self._dbuser = dbuser if dbuser is not None else config['database'].get('username')
+            assert self._dbuser
+
+            self._dbpass = dbpassword if dbpassword is not None else \
+                config['database'].get('password')
+            assert self._dbpass
+
+        # Create Database Connections Pool
+        try:
+            self._dbpool = psycopg2.pool.SimpleConnectionPool(5, 10, host=self._dbhost,
+                                                              port=self._dbport,
+                                                              dbname=self._dbname,
+                                                              user=self._dbuser,
+                                                              password=self._dbpass)
+
+            if self._dbpool:
+                log.info("SQL Database Connected.")
+        except (Exception, psycopg2.DatabaseError) as error:
+            log.warning("Error connecting to SQL database.", error)
```

Comment:
fatal errors should not be suppressed. Responsibility for handling a failed connection should be differed to the caller.

Reply:
Its worth noting that this line lead to downstream errors during testing where the database fails to connect. Granted in this situation any subsequent calls fail but they do so for the wrong reason.
```
test setup failure
test_dbm = <database.database_manager.DatabaseManager object at 0x7fb37a42ef60>

    @pytest.fixture()
    def test_dbm_pool(test_dbm) -> psycopg2.pool.SimpleConnectionPool:
        """
        Test fixture for Database Manager's connection pool.
    
        A DATABASE CONFIGURATION AND CONNECTION IS REQUIRED FOR THESE TESTS.
        """
>       return test_dbm._dbpool
E       AttributeError: 'DatabaseManager' object has no attribute '_dbpool'

tests/test_database_manager.py:40: AttributeError
```

Reply:
unless Github's diff is deceiving me i don't see a `raise` statement in here. the exception raised here should not be suppressed.

Reply:
with the log.warning on line 58, this is the output for a captured DatabaseError:

`E           psycopg2.DatabaseError: ('Unable to connect to database.', OperationalError('could not connect to server: Conne
ction refused (0x0000274D/10061)\n\tIs the server running on host "localhost" (::1) and accepting\n\tTCP/IP connections on
port 5432?\ncould not connect to server: Connection refused (0x0000274D/10061)\n\tIs the server running on host "localhost"
 (127.0.0.1) and accepting\n\tTCP/IP connections on port 5432?\n'))`

No other errors are directly captured, and will raise.